### ISSUE TITLE
Feature Request: Standard Opacity options like the traditional console. 

### ISSUE THREAD

**Issue by NOFUNEVER** [microsoft/terminal#603]
Perhaps I overlooked it browsing through the json file but i didn't see another means to achieve transparency and i'm not a fan of the acrylic effects blurring factor or the fact that it turns off when the window is not selected. The traditional console provides for a static opacity just like in most nix envs, i'd hope it's just a matter of time before it's implemented here?

**Comment by DHowett-MSFT** [comment]
It's not supported right now, but it's definitely something we don't want to lose track of. We'll need to do a bit of heavy lifting to get the composition visuals set up right, but it is "in plan."

**Comment by zadjii-msft** [comment]
See also #593, which has additional info.

**Comment by silverqx** [comment]
I like classic transparency, but never used acrylic, one of the must-have feature for me :)

**Comment by TCB13** [comment]
+1 on this!

**Comment by zadjii-msft** [comment]
Please do not "+1" issues, it creates unnecessary noise. There's a perfectly good +1 button right here:
![image](https://user-images.githubusercontent.com/18356694/60544971-8b497d00-9cdf-11e9-88ca-519566f7ad44.png)

**Comment by TCB13** [comment]
@zadjii-msft with microsoft there is no such thing as a "perfectly good +1 button", even when making noise nobody really cares or listens to anything. I guess anyone who has ever used any Microsoft product knows this.

**Comment by DHowett-MSFT** [comment]
@TCB13 Just so you know, we're all actual people here sitting in our offices looking at our inboxes, and our triage queue, and the github issues list. We're reading each one of these issues, and we _feel things_ about them and the things people say in them. Please try to avoid being mean.

**Comment by TCB13** [comment]
@DHowett-MSFT 

The fact that acrylic was included instead of opacity really means that nobody actually considered the usability of this terminal. Looks to me that some PM just said "oh transparent things look cool, lets make it even cooler and add a blur to it!". Terminal apps / windows aren't transparent in other implementations because it looks great, they are transparent because it allows you to see other stuff behind the window - an useful feature as most used will agree.

Maybe you should ask the other actual people in your company why I was so aggressive in my other comment. :)

**Comment by mdtauk** [comment]
@TCB13 there were voices asking for Acrylic to be added to the Command terminal, and at the time it was not possible for backwards compatibility reasons, but a new Terminal makes that possible.

The issue with this new version is that Opacity is not immediately possible with the new Windowing APIs which this is using, and is something the team is aware of, and will try to make possible.

There is no need to be so dismissive, just because it is not a feature you personally want or need.

**Comment by iCodeSometime** [comment]
This would be very useful.

In addition to what's already been said, the ability to have the window acrylic when focused and transparent when unfocused would be very nice as it would make that transition less jarring.

**Comment by lkspencer** [comment]
agreed. I personally want the regular opacity setting because I like very dark backgrounds with a slight transparency. When I tried using the acrylic options it made the background too bright for my liking. I'm a vampire I guess 🤷‍♂.

Glad to hear that it's at least planned. Thanks for making this tool and I look forward to future releases with more features like this one.

**Comment by Oneliey** [comment]
agreed. I personally think that the window turns straight to solid color suddenly when it loses focus. Traditional transparency is a better transition
Glad to hear that it's at least planned. Thanks for making this tool and I look forward to future releases with more features like this one.

**Comment by silverqx** [comment]
The true is, that I will not use console, when there will not be classic transparency :), it's one must to have feature for me :)

**Comment by ratatoeey** [comment]
<details>
<summary>CLICK ME</summary>
<p>

this is not related to this issue but rather to a set of posts.

> There is no need to be so dismissive, just because it is not a feature you personally want or need.

when people are seem to be way out of touch with the common expectations of majority of app users, they get passionate responses.

and transparency has been available for a long time, so its not a single person pressuring the devs to please one single person. as your last statement implies, for some reason; feel free to edit it.

maybe TCB13 was rude or aggressive in some other thread or occassion? as its mostly harmless hyperbole in https://github.com/microsoft/terminal/issues/603#issuecomment-508031247

---

also, silverqx, see https://github.com/microsoft/terminal/issues/603#issuecomment-546613996 https://github.com/microsoft/terminal/issues/603#issuecomment-507248317 and how long it has been since https://github.com/microsoft/terminal/tags?after=RS2-final been since that first public release. It seems, it will take a long time.

don't do it twice, see https://github.com/microsoft/terminal/issues/603#issuecomment-507835880

---

from https://github.com/microsoft/terminal/issues/603#issuecomment-529696036,


</p>
</details>

**Comment by zadjii-msft** [comment]
So for the record, I've played with this. A rudimentary implementation isn't super tricky, but it's not... polished. The entire window becomes equally transparent including XAML controls (tab row, dialogs, etc.). I think from an architecture standpoint, it would be exceptionally tricky to try and get _just_ the "Terminal" content transparent, and even then, all Panes would be equally transparent, including separators, and dialogs would also still be transparent.

![image](https://user-images.githubusercontent.com/18356694/68593349-daa47480-045a-11ea-9ad7-3214a445ade5.png)

Weirdly enough, the MenuFlyout for the new tab dropdown _doesn't_ become transparent, which raises more questions.
![image](https://user-images.githubusercontent.com/18356694/68593407-0293d800-045b-11ea-9ebe-ba311237c32b.png)

Honestly in _my_ opinion, the experience feels a little unpolished. If this is what people really want, I'm not going to say no, but I also want to make sure we're shipping something high quality. Hence why I'm leaving it on the backlog, to try and find a _better_ solution.

**Comment by heaths** [comment]
Actually, I like the non-client regions transparent. That's how cmd works, and has for a while (maybe RS5?):

![image](https://user-images.githubusercontent.com/1532486/68614416-b1dda880-0475-11ea-953d-be6f6bc58c3e.png)

**Comment by Brassfield** [comment]
> Honestly in _my_ opinion, the experience feels a little unpolished. If this is what people really want, I'm not going to say no, but I also want to make sure we're shipping something high quality. Hence why I'm leaving it on the backlog, to try and find a _better_ solution.

I quite like this effect - the only solid windows I want are dialogs & dropdowns - now I suppose I just  need to build this from source :-/

Thanks for linking to the modification!

**Comment by fhelwanger** [comment]
Yeah, I also think that everything transparent is a feature and not a bug 😄 

The way cmd does it is very useful because you can read what is beneath (of course it depends on the transparency level to looks "good").

**Comment by admered1** [comment]
I would strongly suggest that the devs look at the functionality of gnome-terminal wrt transparency as a guide. For my part, I don't have a use for Acrylic and its fuzziness. I want it transparent so I can see through it, whether it's focussed or not. Thanks.

**Comment by silverqx** [comment]
> So for the record, I've played with this. A rudimentary implementation isn't super tricky, but it's not... polished. The entire window becomes equally transparent including XAML controls (tab row, dialogs, etc.). I think from an architecture standpoint, it would be exceptionally tricky to try and get _just_ the "Terminal" content transparent, and even then, all Panes would be equally transparent, including separators, and dialogs would also still be transparent.

It's ok to have the tab bar transparent, the whole window has to be transparent, with scrollbars too.
Much better would be to have little more transparency for terminal content, than for tab bar and statusbar, eg about 10%, but this has lower priority and it's not substantial.
Modal dialogs have to be opaque, nobody wants to have transparent dialogs. 🙂

The essential thing here is to make it look **good**. 🚀

> Weirdly enough, the MenuFlyout for the new tab dropdown _doesn't_ become transparent, which raises more questions.

This is ok, this dropdown or flyout menus don't have to be transparent, it's much better to have them opaque.

> Honestly in _my_ opinion, the experience feels a little unpolished. If this is what people really want, I'm not going to say no, but I also want to make sure we're shipping something high quality. Hence why I'm leaving it on the backlog, to try and find a _better_ solution.

You have to little play with it, purple background is not good example for a transparent background, it looks much better with black background.
Some examples [1](https://i.stack.imgur.com/bjYQ1.png), [2](https://fedoramagazine.org/wp-content/uploads/2014/05/Screenshot-from-2014-05-15-173413.png).

If you set up good and balanced transparency for a terminal, it can help you in some scenarios, in my workflow it helps me in 5-10% cases, when I don't have to switch to the underlying window. This is good from a practical point of view and added value is that it looks good. 🙂

Here is a [link](https://github.com/Maximus5/ConEmu/blob/be9695f5b565c28379237a5fdea3abab7fea3e3a/src/ConEmu/ConEmu.cpp#L12824) to code, how it is implemented in conemu.

**Comment by ackvf** [comment]
@zadjii-msft I would be totally fine with your experimental transparency if it were implemented as in my suggestion
focused - useAcrylic
unfocused - use transparency

https://github.com/microsoft/terminal/issues/4413

**Comment by effolkronium** [comment]
Can u please add.a different options for transparent background and text(foregorund)

**Comment by Nirmal4G** [comment]
@zadjii-msft @cinnamon-msft 

I'd like a **Peek** button on the TitleBar near the new tab **`+`** button or the minimize **`-`** button, just like Aero Peek.

If you guys are redoing the composition work then please consider adding it to the TitleBar. So that when I hover over it, I can see the screen behind, rather than staying transparent all the time.

__Relatively Related__

Can we have a Fullscreen button on the TitleBar on all apps that allow maximize (yes in terminal too)! I think we should add **Peek** to the list too IMHO, if we're adding in the Windowing system itself.

**Comment by zadjii-msft** [comment]
@Nirmal4G I've moved your first request to #5426. I'm pretty sure there's nothing we can do about your second request - that sounds like a pretty extensive change to the windowing of _all_ Windows applications. It might be a request that would fit in well over at [Microsoft/PowerToys](https://github.com/microsoft/PowerToys). IIRC, one of the original powertoys worked by modifying the titlebar caption buttons 
![image](https://user-images.githubusercontent.com/18356694/79763399-a7a7bc00-82e9-11ea-8d52-740a57887da6.png)

**Comment by glowtape** [comment]
Since this seems to be the catch-all issue for transparency and acrylic... I just installed this Notepads App that some other MSFT guy did in his spare time, and it seems to be able to maintain acrylic transparency while unfocused in background. It's using XAML and Windows.UI, too.

Please add this capability, too, at some point. Thanks.

The project page: https://github.com/JasonStein/Notepads

**Comment by iCodeSometime** [comment]
I don't think they're saying they *can't* keep it acrylic when unfocused, they're saying they've decided not to as a broad company wide design decision in order to save battery life.

**Comment by glowtape** [comment]
Yeaaaa, well... Transparency/acrylic in the terminal is the sort of eye candy that the user probably expects to be there all the time. Otherwise no one would want it, because having it while it's focused doesn't do that much in that regard.

Also, terminal users are likely advanced users, also considering actually needing to edit a JSON to enable it, and would probably be aware of the battery impact in some fashion.

Could still make it optional.

**Comment by zadjii-msft** [comment]
@ future self:

* https://github.com/JasonStein/Notepads/blob/58530f19dd0173bab13e40c9511e5277e42c4129/src/Notepads/Brushes/HostBackdropAcrylicBrush.cs
  - Shoot, this is using Win2D, which I'm not sure we can use, since it's C#
  - https://github.com/microsoft/Win2D/blob/master/winrt/docsrc/CanvasDevice.xml / https://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_CanvasDevice_GetSharedDevice.htm
* There is however https://docs.microsoft.com/en-us/uwp/api/windows.ui.composition.compositor.createhostbackdropbrush?view=winrt-18362
  - this looks promising
* https://stackoverflow.com/questions/43208841/how-to-use-acrylic-accent-createhostbackdropbrush-in-windows-10-creators-upd/44576160
* Then we'll write some code
```c++

            auto rootVisual = winrt::Windows::UI::Xaml::Hosting::ElementCompositionPreview::GetElementVisual(RootGrid());
            auto compositor = rootVisual.Compositor();
            auto rootSprite = compositor.CreateSpriteVisual();
            rootSprite.Size(winrt::Windows::Foundation::Size(
                ::base::saturated_cast<float>(RootGrid().ActualWidth()),
                ::base::saturated_cast<float>(RootGrid().ActualHeight())));

            auto b = rootVisual.Compositor().CreateHostBackdropBrush();

            rootSprite.Brush(b);
            winrt::Windows::UI::Xaml::Hosting::ElementCompositionPreview::SetElementChildVisual(RootGrid(), rootSprite);
```
* Oh no, even with this, we get pre-blurred content 
![image](https://user-images.githubusercontent.com/18356694/81589666-38206c00-937f-11ea-8884-c234d065dc1b.png)

* Which leads me to https://github.com/Microsoft/WindowsCompositionSamples/issues/202
* Which brings me full circle to the investigation I already did on this last year https://github.com/microsoft/terminal/issues/1753#issuecomment-508070516


I'm not promising anything here, just leaving notes from the tab's I've got open this morning

**Comment by Varstahl** [comment]
Wait, did I read it correctly? Having a partially transparent window or an acrylic without blur is not doable because of security issues? *What?*

I, like many, quite desperately want/need a terminal semi-transparency like it's been in use on the *nix front in ages. Do we really have to go the "screensaver route", capture the screen during painting, paint the capture as a background image in the window, and then paint the opacity on top of it?

I mean, how feasible could a semi-transparent window be a vector for an attack… even a fully transparent keylogger would deactivate the window. Genuinely curious. And baffled. But mostly curious.

**Comment by zadjii-msft** [comment]
Granted, that thread's pretty old, from a time when the only way to do Acrylic was from a pure UWP application. UWPs are pretty highly insulated against being able to read the state of other processes on the system. If an app in this context could trivially ready the contents of the window behind it, then it could theoretically read the contents of anything else that's running on the system, in _any_ context. This is something that's absolutely not possible for UWP apps.

The world has changed a lot since then - there's apps like the Terminal that use UWP XAML (and acrylic) but _aren't_ UWP applications. In this new model, it might be possible for us to do something differently, we just need to do more research.

**Comment by tajetaje** [comment]
Reading through the [source for the AcrylicBrush](https://github.com/microsoft/microsoft-ui-xaml/tree/master/dev/Materials/Acrylic), I may have found a workaround. But do take this with a grain of salt as I have minimal experience with windows UI development.
It looks like a [factory method](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Materials/Acrylic/AcrylicBrush.cpp#L798) in the Acrylic Brush has support for pre-blurred backgrounds, seemingly where the shell would have already done it, IF that method is exposed and IF a non-blurred background is passed as a pre-blurred background and IF I'm not totally off the mark here, the factory might not add a blur effect whatsoever. But again, I have a high chance of being 150% wrong here so, take that as you will.

**Comment by alxkvx** [comment]
OMG, this discussion is more than an year old, but essential Glass Transparency is still not implemented. It is standard for unix terminals.

**Comment by zadjii-msft** [comment]
@alxkvx As discussed in this thread at great length, this is a technically hard problem, hence why it still hasn't been implemented. If you've got any suggestions as to _how_ this could be implemented with our UI stack, I'd be happy to hear them.

**Comment by alxkvx** [comment]
> @alxkvx As discussed in this thread at great length, this is a technically hard problem, hence why it still hasn't been implemented. If you've got any suggestions as to _how_ this could be implemented with our UI stack, I'd be happy to hear them.

well, i didn't dig into the problem actually, And i dont understand why UI stack has problems with that, i know that using WSL terminal i can just right click > Properties > Set Opacity and i will get glass transparency. For me personally this is essential feature.

**Comment by taylus** [comment]
The blur and going opaque when the terminal isn't focused is just silly. Who asked for this? 🤷 I'll keep using cmd and powershell.

**Comment by tajetaje** [comment]
> > @alxkvx As discussed in this thread at great length, this is a technically hard problem, hence why it still hasn't been implemented. If you've got any suggestions as to _how_ this could be implemented with our UI stack, I'd be happy to hear them.
> 
> well, i didn't dig into the problem actually, And i dont understand why UI stack has problems with that, i know that using WSL terminal i can just right click > Properties > Set Opacity and i will get glass transparency. For me personally this is essential feature.

The issue is that the UI framework (UWP) that Windows terminal is built in simply does not support full transparency like win32 apps (e.g. cmd). The only tool they have access to is Acrylic which is meant less for transparency and more as a design accent. Implementing it under UWP would take a hacky solution like my above comment or a total reworking of how the console is rendered (at least that's my understanding, windows UI isn't my field of development)

**Comment by mdtauk** [comment]
> > > @alxkvx As discussed in this thread at great length, this is a technically hard problem, hence why it still hasn't been implemented. If you've got any suggestions as to _how_ this could be implemented with our UI stack, I'd be happy to hear them.
> > 
> > 
> > well, i didn't dig into the problem actually, And i dont understand why UI stack has problems with that, i know that using WSL terminal i can just right click > Properties > Set Opacity and i will get glass transparency. For me personally this is essential feature.
> 
> The issue is that the UI framework (XAML) that Windows terminal is built in simply does not support full transparency like win32 apps (e.g. wsl terminal). The only tool they have access to is Acrylic which is meant less for transparency and more as a design accent. Implementing it under their UI stack would take a hacky solution like my above comment or a total reworking of how the console is rendered (at least that's my understanding, windows UI isn't my field of development)

Xaml does support transparency, but UWP windows and Xaml Islands does not.

When Terminal is able to move over to WinUI 3 Desktop, and can tap into the HWND itself, then it will be able to implement a non-blurred Transparency.

Acrylic itself will be delayed for WinUI 3 - as they have to re-work it whilst extracting it from the OS

**Comment by zadjii-msft** [comment]
Wait no, lets clear up some misconceptions - We're already able to access the HWND directly, because we're already a win32 app. We're working with the XAML and composition teams to try and figure out a solution to this. From my understanding (currently), the XAML Island always has an opaque background, which means that we can't just have transparent components in XAML that are transparent all the way through to the HWND. I have no idea if WinUI 3 will magically solve this for us, which is something we'll need to discuss more with that team. Fortunately, they work in the hallway next to us (or at least they did when we all worked in office buildings), so having these discussions isn't too hard 😄

**Comment by AnuthaDev** [comment]
> @ future self:
> 
> * https://github.com/JasonStein/Notepads/blob/58530f19dd0173bab13e40c9511e5277e42c4129/src/Notepads/Brushes/HostBackdropAcrylicBrush.cs
>   
>   * Shoot, this is using Win2D, which I'm not sure we can use, since it's C#

I am pretty sure Win2D supports c++/winrt (its written in c++!), but even if not for this use case there is this:

https://github.com/fobrs/Win2DinMFC

Also, I believe custom acrylic should be possible because I was able to achieve acrylic (with [rounded](https://github.com/windows-toolkit/Microsoft.Toolkit.Win32/issues/168) corners) in WPF , by following this sample:

https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/tree/a59e7586c0bd1a967e1e25f6ca0363e20151afe5/dotnet/WPF/AcrylicEffect

And this post is phenomenal:

https://notes.yvt.jp/Desktop-Apps/Enabling-Backdrop-Blur

**Comment by zadjii-msft** [comment]
Those are all really helpful if you're using WPF XAML, but we're using UWP XAML, which will unfortunately always have the opaque backdrop. We're working with the WinUI team to hopefully get that restriction relaxed for WinUI 3.0. There might be a while before there's any other progress on this issue, while we work on the technical specifics with them.

**Comment by AnuthaDev** [comment]
@zadjii-msft So, you are telling me the wpf control of terminal could possibly support acrylic, but not the UWP one...weird times 😅😅

**Comment by zadjii-msft** [comment]
Oh no the UWP XAML one can support acrylic just fine, it's just "traditional opacity" (opacity without the acrylic effect) that the UWP XAML one can't support currently.

**Comment by DHowett** [comment]
I'd love for you to continue this discussion elsewhere (perhaps file a new issue if you have a problem?) because this is the thread for _non-blurry transparency_. We already have "acrylic" in the UWP control, and further discussing "how to get acrylic in the UWP control" is ... I mean, like, trying to explain to a horse how it could become a horse if it really, really wanted to?

**Comment by AnuthaDev** [comment]
@zadjii-msft @DHowett Sorry, I didn't mean that. By acrylic I meant *acrylic with customizable blur radius*

**Comment by AnuthaDev** [comment]
@zadjii-msft  I want to do some experiments, can you point me where the xaml hosting window is created in your code? Its a very large codebase 😅😅

**Comment by greg904** [comment]
@AnuthaDev It's in `src/cascadia/WindowsTerminal/IslandWindow.cpp`

**Comment by AnuthaDev** [comment]
Okay, its atleast possible for wpf for sure:
![Screenshot (363)](https://user-images.githubusercontent.com/36439704/84429049-2adbf300-ac45-11ea-8cb8-0119f328d3aa.png)

**Comment by AnuthaDev** [comment]
Maybe the window in IslandWindow.cpp can be created with WS_EX_NOREDIRECTIONBITMAP and the method followed [here](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/blob/master/dotnet/WPF/AcrylicEffect/AcrylicEffect/CompositionHostControl.xaml.cs) can be used to create an acrylicbrush. Using createhostbackdropbrush() introduces blur automatically, so createbackdropbrush() is the only option. Or maybe it won't work,...  idk. Will try and let you know...

**Comment by tajetaje** [comment]
> Maybe the window in IslandWindow.cpp can be created with WS_EX_NOREDIRECTIONBITMAP and the method followed [here](https://github.com/microsoft/Windows.UI.Composition-Win32-Samples/blob/master/dotnet/WPF/AcrylicEffect/AcrylicEffect/CompositionHostControl.xaml.cs) can be used to create a _non blur_ acrylicbrush. Using createhostbackdropbrush() introduces blur automatically, so createbackdropbrush() is the only option. Or maybe it won't work,... idk. Will try and let you know...

I don’t think so as that method is under WPF. XAML is used in two different frameworks WPF (that method) and UWP (used in WT).  I’ve poked around in the source for UWP acrylic, and the only thing that could possibly get traditional transparency working is a really Hakki workaround where are you basically trick the OS into thinking that the background has already been blurred so it doesn’t have to add a blur or later.

**Comment by AnuthaDev** [comment]
@zadjii-msft @DHowett Okay, so here is what I found:
It is certainly *possible* to get a custom blur acrylic in a c++ win32 app using win2d:
![Screenshot (364)](https://user-images.githubusercontent.com/36439704/84570100-938fb080-ada8-11ea-9893-15aab24d37b4.png)

HOWEVER!! You cannot do so with xaml islands. As you already know there will definitely be an opaque background behind xaml...

To do this we need to use composition apis and render to a hwnd.

Therefore, as it currently stands, to get non blur acrylic we would need to remove xaml islands and use [this](https://docs.microsoft.com/en-us/windows/uwp/composition/composition-native-interop) instead of a swapchainpanel.

(If you already know this, sorry for wasting your time...)

**Comment by AnuthaDev** [comment]
Hence, no transparency without a significant architecture change.

(A conclusion that was already apparent and to which I contributed absolutely nothing😅😅)

**Comment by tajetaje** [comment]
Yeah, so TLDR for anyone who doesn’t want to read all of that ^ is basically that the framework that is used for windows terminal does not currently support this feature HOWEVER development on that framework is ongoing and transparency (as far as I understand it) is in the works. So only once the framework (XAML islands) supports transparency can this issue be started on.

**Comment by alxkvx** [comment]
> Yeah, so TLDR for anyone who doesn’t want to read all of the previous discussion is basically that the framework that is used for windows terminal does not currently support this feature HOWEVER development on that framework is ongoing and transparency (as far as I understand it) is in the works. So only once the framework (XAML islands) supports transparency can this issue be started on.

i wonder why transparency feature was not included initially in that project and was not considered when UI engine was chosen. Having 10+ years working with Linux terminals, all have that feature and it is actively used by many users. Strange for me.

**Comment by 50l3r** [comment]
Powershell & CMD have the option to set transparency

**Comment by alxkvx** [comment]
> Powershell & CMD have the option to set transparency. I understand that technologies used are different but many users use transparency configs

yea, the same as WSL terminal

**Comment by Varstahl** [comment]
The CMD and PWSH transparency is achievable with this terminal, but from what I gather the majority of people (myself included) don't want *that version* of transparency, rather a *nix-like terminal transparency, where only the background is translucent instead of everything, text included.

Also, there are some hackish ways to fake transparency, with a screen RECT grab, painted over the terminal itself, and then painting a semi transparent color over, but even in that there are limitations and pitfalls.

**Comment by AnuthaDev** [comment]
Maybe this can be implemented by creating a window below the xaml islands and setting WS_CLIPSIBLINGS on it, then composition api and directx [interop](https://docs.microsoft.com/en-us/windows/uwp/composition/composition-native-interop) can be used to render content with translucent background to this window. So, you wouldn't need to remove xaml islands and stuff like scrollbars should still work, just the swapchainpanel part would be replaced. Or maybe if that doesn't work, you could cut a hole over the swapchainpanel part using HRGN, so the composition window below it becomes visible. There shouldn't be any noticeable performance regression moving from swapchainpanel to a hwnd (probably)

**Comment by tajetaje** [comment]
The problem with pursuing a hackey workaround is that it is hackey and inherently prone to bugs. WinUI 3 will most likely support the proposed functionality so there are plans for implementing it, it’s just a waiting game for the official tool chain to support it. The devs have already confirmed that they are directly collaborating with the WinUI team on this.

**Comment by zadjii-msft** [comment]
> (A conclusion that was already apparent and to which I contributed absolutely nothing😅😅)

I wouldn't say that you contributed nothing - I'm always happy to have external confirmation of my own research. I would have been even happier if you had proven me wrong and found an effective way to do this 😉 

As mentioned before in this thread - We're working with the WinUI team to get this added in to WinUI 3.0. I believe this is being tracked in https://github.com/microsoft/microsoft-ui-xaml/issues/1247. This is the path we'll be pursuing to get this feature added to the Terminal, because driving this solution also means driving an important dev platform feature for the entire platform, one that'll help improve other applications on Windows as well.

**Comment by tajetaje** [comment]
> As mentioned before in this thread - We're working with the WinUI team to get this added in to WinUI 3.0. I believe this is being tracked in [microsoft/microsoft-ui-xaml#1247](https://github.com/microsoft/microsoft-ui-xaml/issues/1247). This is the path we'll be pursuing to get this feature added to the Terminal, because driving this solution also means driving an important dev platform feature for the entire platform, one that'll help improve other applications on Windows as well.

Since the terminal team has chosen a direction on this one, should the "help wanted" tag be removed?

**Comment by Chiramisu** [comment]
> > (A conclusion that was already apparent and to which I contributed absolutely nothing😅😅)
> 
> I wouldn't say that you contributed nothing - I'm always happy to have external confirmation of my own research. I would have been even happier if you had proven me wrong and found an effective way to do this 😉
> 
> As mentioned before in this thread - We're working with the WinUI team to get this added in to WinUI 3.0. I believe this is being tracked in [microsoft/microsoft-ui-xaml#1247](https://github.com/microsoft/microsoft-ui-xaml/issues/1247). This is the path we'll be pursuing to get this feature added to the Terminal, because driving this solution also means driving an important dev platform feature for the entire platform, one that'll help improve other applications on Windows as well.

@zadjii-msft, I can see how WinUI would play a part in this, but not sure how the borderless issue gets us there. In terms of the actual result most people are after, I think this is a beautiful rendition from @mdtauk ([#743](https://github.com/microsoft/terminal/pull/743#issuecomment-492019264)) of what could be:

> ![image](https://user-images.githubusercontent.com/7389110/57660081-c9c69380-75dc-11e9-9dc6-a45942a24307.png)

**Comment by tajetaje** [comment]
> > > (A conclusion that was already apparent and to which I contributed absolutely nothing😅😅)
> > 
> > 
> > I wouldn't say that you contributed nothing - I'm always happy to have external confirmation of my own research. I would have been even happier if you had proven me wrong and found an effective way to do this 😉
> > As mentioned before in this thread - We're working with the WinUI team to get this added in to WinUI 3.0. I believe this is being tracked in [microsoft/microsoft-ui-xaml#1247](https://github.com/microsoft/microsoft-ui-xaml/issues/1247). This is the path we'll be pursuing to get this feature added to the Terminal, because driving this solution also means driving an important dev platform feature for the entire platform, one that'll help improve other applications on Windows as well.
> 
> @zadjii-msft, I can see how WinUI would play a part in this, but not sure how the borderless issue gets us there. In terms of the actual result most people are after, I think this is a beautiful rendition from @mdtauk ([#743](https://github.com/microsoft/terminal/pull/743#issuecomment-492019264)) of what could be:
> 
> > ![image](https://user-images.githubusercontent.com/7389110/57660081-c9c69380-75dc-11e9-9dc6-a45942a24307.png)

https://github.com/microsoft/microsoft-ui-xaml/issues/1247 tracks both borderless and transparency

**Comment by landonth** [comment]
> Wait no, lets clear up some misconceptions - We're already able to access the HWND directly, because we're already a win32 app. We're working with the XAML and composition teams to try and figure out a solution to this. From my understanding (currently), the XAML Island always has an opaque background, which means that we can't just have transparent components in XAML that are transparent all the way through to the HWND. I have no idea if WinUI 3 will magically solve this for us, which is something we'll need to discuss more with that team. Fortunately, they work in the hallway next to us (or at least they did when we all worked in office buildings), so having these discussions isn't too hard 😄

@zadjii-msft great to hear about this sort of inter-team communication. just goes to show how sociological (not just technological) building software really is. thank you for your continued work and navigating all the (sometimes very... passionate 😅) community feedback on this. 💞

**Comment by desmap** [comment]
> the majority of people (myself included) don't want that version of transparency, rather a *nix-like terminal transparency, where only the background is translucent instead of everything, text included.

I prefer full transparency because it gives a more consistent look and feel. Giving some elements transparency and some not looks less clean. The opacity level has to be of course right and then, things look really gorgeous. If you can change opacity on the fly things get even better, we did this in the other issue with AHK. I posted following screenshots to the other issue, here with some additional window behind just to show its visual power:

With a browser behind __(pls zoom in to the native resolution to check the readability of both the vim welcome message and the Github issue in Chrome behind)__:

![image](https://user-images.githubusercontent.com/43666255/108164064-62623f80-70f0-11eb-80d1-0db881c3d1d1.png)

Just the wallpaper behind:

![image](https://user-images.githubusercontent.com/43666255/108164122-7c038700-70f0-11eb-8969-d16ec2dad429.png)

As said, if you set the opacity too low/the transparency too high, of course things will look messed up.

**Comment by darkphase** [comment]
How alacritty is doing that
https://github.com/alacritty/alacritty

**Comment by denodaeus** [comment]
So basically, beyond the above hack that sometimes works with AHK, this issue has been open for 2 years, is one of the most often requested features in Issues that constantly get closed because of duplicates, gets handed off to a different team to implement of which is in the "Freezer" part of their backlog, so this feature probably isn't going to happen for another 3 years.

I've been tracking this feature now for more than a year, and it really is the single make it or break it feature keeping me from moving to terminal over Hyper.js -- but just seeing the back and forth between teams at Microsoft and now seeing this basically buried over on the microsoft-ui-xaml side for 2 years is incredibly dissapointing.

**Comment by mdtauk** [comment]
I think this requires WinUI 3 to be done before Terminal can have full ownership of the window transparency, but I may be mistaken.

One issue with full transparency, is accessibility concerns with contrast ratios.  If the text could remain fully opaque, whilst background colours and window frames become translucent, this would be a better approach, and then transparency could be disabled with high contrast modes.

**Comment by tajetaje** [comment]
> I think this requires WinUI 3 to be done before Terminal can have full ownership of the window transparency, but I may be mistaken.
> 
> One issue with full transparency, is accessibility concerns with contrast ratios. If the text could remain fully opaque, whilst background colours and window frames become translucent, this would be a better approach, and then transparency could be disabled with high contrast modes.

Honestly with a developer app like terminal I think we can trust users to know that transparency can conflict (depending on config it can be fine) with high contrast. There are times that the UI/UX should restrict potentially bad experiences, but developer features are rarely the place to limit user freedom.

**Comment by iCodeSometime** [comment]
Agreed. Sounds like you're worried someone might _choose_ to set their transparency to a setting that your team personally doesn't like.

This is a feature of every major terminal, and literally no one has a problem with this. Just do a few quick google searches
- "I can't read my terminal"
- "Terminal text too faint"
- "Terminal too transparent"

There are none.

**Comment by tajetaje** [comment]
> Agreed. Sounds like you're worried someone might _choose_ to set their transparency to a setting that your team personally doesn't like.
> 
> This is a feature of every major terminal, and literally no one has a problem with this. Just do a few quick google searches
> 
> * "I can't read my terminal"
> * "Terminal text too faint"
> * "Terminal too transparent"
> 
> There are none.

a little harsh, but nonetheless yeah that's what I was getting at. On the other hand if WT does become the default terminal some kind of accomidation might be a good idea. My commend only applies if WT remains a dev/power user tool

**Comment by orcmid** [comment]
> @tajetaje My commend only applies if WT remains a dev/power user tool

Really, so no developers requiring assistance need apply?  Power users requiring assistance technologies can stay away?

Or are we talking about user-driven adjustments that the producer of the code feeding the terminal display need be concerned about?

**Comment by tajetaje** [comment]
> > @tajetaje My commend only applies if WT remains a dev/power user tool
> 
> Really, so no developers requiring assistance need apply? Power users requiring assistance technologies can stay away?
> 
> Or are we talking about user-driven adjustments that the producer of the code feeding the terminal display need not be concerned about?

No absolutely not, I more mean that I trust developers and power suers to know that turning on a transparency effect where you with have light text on light background is a bad idea AND I trust them to know how to either tune the settings or turn it off. IF terminal does become a part of the Windows OS then the paradigm is different and the WT should and likely will follow other apps and remove some user control in favor of UX.

**Comment by W4RH4WK** [comment]
Just have sane defaults. Don't have the transparency on by default. If a user enables it, they know where to disable it if it causes problems. I doubt such an option would _clutter_ the UI unnecessarily, but _even_ if that turns out to be a problem, let's just keep transparency as a config option. Power users still have access to it that way. It should be implemented as it is a pretty common feature in terminal emulators.

**Comment by zadjii** [comment]
Nobody is arguing that this shouldn't be implemented. It will be, when we can implement it in a sane, polished way befitting the default terminal experience in Windows. 

Until that time, I'm hoping we can all just remain patient. We want this just as badly as y'all, we're just limited to the engineering constraints common in our industry. We've got a small dev team and a LOT of projects in motion.

**Comment by yumetodo** [comment]
We should know the difference between "has an ability" and "is default".

**Comment by zerols** [comment]
like alacritty how to achieve this function？

**Comment by zadjii-msft** [comment]
Ho boy I hate bumping a thread with this many folks on it but I gotta keep my notes somewhere. Just so happens this is the place. No spoilers here this week.

* [ ] Ship the updated TTH package.
* [ ] Fix Mouse wheeling opacity
* [ ] Figure out if we can get rid of the titlebar in focus mode
* [ ] do we need `WS_EX_LAYERED` in `AlwaysOnTop` mode? Hopefully not, cause that does weird things each settings reload
* [ ] Re-add the Pane border
* [ ] Add the SUI background
* [ ] change opacity to a 0-100 value, with a `FloatAsIntPercentConversionTrait`
* [ ] `INHERITABLE_SETTING(int, Opacity, UseAcrylic() ? 50 : 100); // Sorry lhecker`
* [ ] update docs

**Comment by silverqx** [comment]
There were so many cries that support from XAML is needed, because XAML doesn't support it, and from the patch is clearly visible that `Background="Transparent"` was supported a long time and `Opacity` too.

Anyway, I'm really happy about this PR that is finally available (soon), many thanks to @zadjii-msft for great work. 🔥🚀
I can't wait to start using it daily. 😍

**Comment by DHowett** [comment]
> from the patch is clearly visible that `Background="Transparent"` was supported a long time and `Opacity` too.

You may have missed that we had to use a private API to get this working. Everything looks easy through that lens.

Without using that API, XAML enforces the presence of a black background under all controls as an "emergency" measure to *explicitly prevent transparency*.

**Comment by silverqx** [comment]
> 
> You may have missed that we had to use a private API to get this working. Everything looks easy through that lens.

I'm pretty sure that this was 1 month work, nothing easy.

> 
> Without us using that API, XAML enforces the presence of a black background under all controls as an "emergency" measure to _explicitly prevent transparency_.

That is absolutely normal that some hacks had to be used, like any other hard PRs, in coding nothing is easy and if you are adding something bigger then you can be pretty sure that you hit some wall and need to use some workarounds.

**Comment by DHowett** [comment]
@quental this will be released in a stable version on our normal "preview -> stable" schedule (so, approximately 6 weeks from now.)

**Comment by tajetaje** [comment]
> On what it depends? ok, thx for the reply, I think it is a great step forward, 1.12 is amazing thx

see #11180

**Comment by DHowett** [comment]
It depends on a new version of Windows. It will be distributed through Windows.
If we can get it backported to Windows 10, it will be released through Windows Update.

**Comment by silverqx** [comment]
Ok, thank you, I was only curious, it doesn't matter where is from the patch, somewhere in the windows core maybe 🙂, or somewhere.

**Comment by ChrTall** [comment]
@DHowett Can you tell us how likely it is, that you will be able to backport it? 
Is this something you can influence or decided by upstream?
Sadly my 7th gen Intel is not supported by Windows 11 and probably won´t be in the future.

**Comment by zadjii-msft** [comment]
> Can you tell us how likely it is, that you will be able to backport it?

Nope. We're working through it right now. We're putting the weight of our team on it, and we'll give updates whenever we have them. Unfortunately, with the holidays coming up, I'd suspect the OS update pipeline to be a little slower than usual, so bear with us.

**Comment by silverqx** [comment]
Does anybody has it on Win10, it is 8 weeks now and I still don't have this opacity feature on Win10.

**Comment by tajetaje** [comment]
See the comment directly above yours. This currently only works in windows 11.

**Comment by silverqx** [comment]
I know that it works on Win11, but from the previous conversation I understood that it will be backported also to Win10 and it will take ~6 weeks. Am I wrongs?

**Comment by tajetaje** [comment]
> I know that it works on Win11, but from the previous conversation I understood that it will be backported also to Win10 and it will take ~6 weeks. Am I wrong?

I don't think anyone ever said 6 weeks (or even that it was definitely possible). All that has been said publicly is that the terminal team is working with the OS people on it and that they would try and make it happen. Read through some of the team members messages above for more info. If it happens it happens and that is great, but if not oh well.

**Comment by zadjii-msft** [comment]
[The comment you're referring to](https://github.com/microsoft/terminal/issues/603#issuecomment-947908313) is discussing the timeframe for the Preview -> Release channel for _the Terminal itself_, not the servicing timeframe. 

With the holidays, we delayed the Preview->Release promotion anyways (there wasn't too much in the last 6 weeks to justify another release).

Additionally, and totally separately, the _internal_ servicing pipleline hasn't ingested this yet. Servicing the OS takes a while - I honestly don't even know how to guess how long that would take.

We'll be sure to update this thread once there's something to share however!

**Comment by silverqx** [comment]
It is not a big deal, because I will upgrade to Win11 in the following few months or weeks, thx for replies, I misunderstood it though and you are referring right comment that confused me. 😁 Thx

**Comment by quental** [comment]
> [The comment you're referring to](https://github.com/microsoft/terminal/issues/603#issuecomment-947908313) is discussing the timeframe for the Preview -> Release channel for _the Terminal itself_, not the servicing timeframe. 
> 
> With the holidays, we delayed the Preview->Release promotion anyways (there wasn't too much in the last 6 weeks to justify another release).
> 
> Additionally, and totally separately, the _internal_ servicing pipleline hasn't ingested this yet. Servicing the OS takes a while - I honestly don't even know how to guess how long that would take.
> 
> We'll be sure to update this thread once there's something to share however!

Since it was delayed, any idea when this will be released (for win11 ofc)? Thanks!

**Comment by zadjii-msft** [comment]
Probably in January some time, but don't hold me to that. Like I said, we don't really want to make a release unless there's actually something to release. There isn't all that much in 1.13 at the moment. We're trying to also line up 1.12 stable with some internal milestones, so I don't think we can actually push 1.12 stable past early Feb.

**Comment by quental** [comment]
> Probably in January some time, but don't hold me to that. Like I said, we don't really want to make a release unless there's actually something to release. There isn't all that much in 1.13 at the moment. We're trying to also line up 1.12 stable with some internal milestones, so I don't think we can actually push 1.12 stable past early Feb.

Thanks for the update! And it's safe to expect this feature in 1.12 stable, right?

**Comment by NobleDraconian** [comment]
@zadjii Will this feature be supported on Windows 10? I just took a look at the preview build that has #11180, but it mentions it's only supported on Windows 11.

**Comment by NobleDraconian** [comment]
@zadjii-msft Will this feature be supported on Windows 10? I just took a look at the preview build that has https://github.com/microsoft/terminal/pull/11180, but it mentions it's only supported on Windows 11.

**Comment by tajetaje** [comment]
> > Can you tell us how likely it is, that you will be able to backport it?
> 
> Nope. We're working through it right now. We're putting the weight of our team on it, and we'll give updates whenever we have them. Unfortunately, with the holidays coming up, I'd suspect the OS update pipeline to be a little slower than usual, so bear with us.

@NobleDraconian see above

**Comment by zadjii-msft** [comment]
As I mentioned above, we're working on servicing the necessary OS side fixes to Windows 10 to make it work, but those take a lot longer than Terminal releases.

**Comment by tajetaje** [comment]
> As I mentioned above, we're working on servicing the necessary OS side fixes to Windows 10 to make it work, but those take a lot longer than Terminal releases.

I know its a bit drastic, but maybe consider locking this issue to collaborators only? There hasn't really been any productive discussion and wont be any more until you have an update.
